### PR TITLE
Projection: `<field>: <1 or true>` and `<field>: <0 or false>`

### PR DESCRIPTION
**Pull request by seeforschauer**

### CODE REVIEW

File: internal/handlers/common/projection.go [hunk 1]
```diff
@@ -19,5 +19,15 @@ import "github.com/FerretDB/FerretDB/internal/types"
 // ProjectDocuments modifies given documents in places according to the given projection.
 func ProjectDocuments(docs []*types.Document, projection *types.Document) error {
 	// TODO
+
+	projectionMap := projection.Map()
```
Comment (seeforschauer) [inline internal/handlers/common/projection.go:23]: it can be done w/o creation of new variable..

File: internal/handlers/jsonb1/msg_count.go [hunk 2]
```diff
@@ -100,9 +100,6 @@ func (s *storage) MsgCount(ctx context.Context, msg *wire.OpMsg) (*wire.OpMsg, e
 		resDocs = append(resDocs, doc)
 	}
 
-	if err = common.SortDocuments(resDocs, sort); err != nil {
-		return nil, err
-	}
```
[the remark this answers is not in the document]
Comment (seeforschauer) [reply internal/handlers/jsonb1/msg_count.go:105]: I'd better won't change it here.

File: internal/handlers/jsonb1/msg_count.go [hunk 3]
```diff
@@ -99,7 +99,6 @@ func (s *storage) MsgCount(ctx context.Context, msg *wire.OpMsg) (*wire.OpMsg, e
 
 		resDocs = append(resDocs, doc)
 	}
-
```
Comment (AlekSi) [inline internal/handlers/jsonb1/msg_count.go:102]: Let's keep a blank line there to separate that block

File: internal/handlers/jsonb1/msg_count.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (AlekSi) [reply internal/handlers/jsonb1/msg_count.go:105]: No, sorting is really not required. But that's not for that PR

File: internal/handlers/jsonb1/msg_count.go [hunk 4]
```diff
@@ -45,7 +45,7 @@ func (s *storage) MsgCount(ctx context.Context, msg *wire.OpMsg) (*wire.OpMsg, e
 		"let",
 	}
 	if err := common.Unimplemented(document, unimplementedFields...); err != nil {
-		return nil, err
```
Comment (AlekSi) [inline internal/handlers/jsonb1/msg_count.go:48]: No, `Unimplemented` returns protocol error that we should return as-is to the client

File: internal/handlers/common/projection.go [hunk 5]
```diff
@@ -18,6 +18,17 @@ import "github.com/FerretDB/FerretDB/internal/types"
...
 // ProjectDocuments modifies given documents in places according to the given projection.
 func ProjectDocuments(docs []*types.Document, projection *types.Document) error {
-	// TODO
+	if projection.Len() == 0 {
+		return nil
+	}
+	projectionMap := projection.Map()
+	for i := 0; i < len(docs); i++ {
+		for k := range docs[i].Map() {
+			if _, ok := projectionMap[k]; !ok {
```
Comment (AlekSi) [inline internal/handlers/common/projection.go:27]: We should check the value too. For example, 0 or false should remove the field.

Or the plan is to do that in a different PR?

Comment (seeforschauer) [reply internal/handlers/common/projection.go:27]: Then I'd add new tests for it too.

Comment (seeforschauer) [reply internal/handlers/common/projection.go:27]: I'd rather add tests and add code to comply with https://docs.mongodb.com/manual/reference/method/db.collection.find/#projection

Comment (AlekSi) [reply internal/handlers/common/projection.go:27]: In this PR, or in different one?

Comment (seeforschauer) [reply internal/handlers/common/projection.go:27]: In this PR

File: internal/handlers/common/projection.go [hunk 6]
```diff
@@ -14,10 +14,47 @@
...
+
+			switch v := v.(type) {
+			case bool:
+				if !v {
+					docs[i].Remove(k)
+					continue
+				}
+				continue
+
+			case int32, int64, float32, float64:
```
Comment (AlekSi) [inline internal/handlers/common/projection.go:48]: float32 is not possible

File: internal/handlers/common/projection.go [hunk 7]
```diff
@@ -14,10 +14,47 @@
...
 // ProjectDocuments modifies given documents in places according to the given projection.
 func ProjectDocuments(docs []*types.Document, projection *types.Document) error {
-	// TODO
+	if projection.Len() == 0 {
+		return nil
+	}
+	projectionMap := projection.Map()
+	for i := 0; i < len(docs); i++ {
+		for k := range docs[i].Map() {
+			if k == "_id" {
```
Comment (AlekSi) [inline internal/handlers/common/projection.go:30]: _id handing is more complicated than that

File: internal/handlers/common/projection.go [hunk 8]
```diff
@@ -14,10 +14,47 @@
...
+				continue
+			}
+
+			v, ok := projectionMap[k]
+			if !ok {
+				docs[i].Remove(k)
```
Comment (AlekSi) [inline internal/handlers/common/projection.go:36]: Do we have a dance test for that? It is not clear from documentation how MongoDB behaves there

File: internal/handlers/common/projection.go [hunk 9]
```diff
@@ -14,10 +14,47 @@
...
+				continue
+			}
+
+			switch v := v.(type) {
+			case bool:
+				if !v {
+					docs[i].Remove(k)
+					continue
+				}
+				continue
```
Comment (AlekSi) [inline internal/handlers/common/projection.go:46]: We don't need both those `continue` statements

File: internal/handlers/common/projection.go [hunk 10]
```diff
@@ -14,10 +14,47 @@
...
+
+			case int32, int64, float32, float64:
+				if compareScalars(v, int32(0)) == equal {
```
Comment (AlekSi) [inline internal/handlers/common/projection.go:49]: Can we use compareNumbers?

File: internal/handlers/common/projection.go [hunk 11]
```diff
@@ -14,10 +14,47 @@
...
+					docs[i].Remove(k)
+					continue
```
Comment (AlekSi) [inline internal/handlers/common/projection.go:51]: We don't need that `continue`

File: internal/handlers/common/projection.go (hunk 10 above)
Comment (seeforschauer) [reply internal/handlers/common/projection.go:49]: then case with int32 , int64 is not handled. `compareScalars` is more convenient here.

File: internal/handlers/common/projection.go (hunk 11 above)
Comment (seeforschauer) [reply internal/handlers/common/projection.go:51]: removed

File: internal/handlers/common/projection.go (hunk 8 above)
Comment (seeforschauer) [reply internal/handlers/common/projection.go]: no, dance has no tests for it at all,
 can it be merged to not to stop Dmitry?

Comment (AlekSi) [reply internal/handlers/common/projection.go]: Sure

**Review by AlekSi — APPROVED** [review]
